### PR TITLE
Compile Whole Program

### PR DESCRIPTION
**Pull request by krisselden**
Compile whole program, instead of per file basis.  Allows external source-maps and declarations to be emitted.

No longer extends from BroccoliFilter and only emits files based on the root files given.

The main export is an approximation of the filter behavior for backwards compat.

Adds a named export for the TypeScript plugin class and a typescript function for creating a broccoli node.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -10,50 +12,96 @@ var getCallerFile = require('get-caller-file');
...
 
 TypeScript.prototype.baseDir = function() {
   return __dirname;
 };
 
+TypeScript.prototype.build = function build() {
+  var inputPath = this.inputPaths[0];
+  var config = ts.parseJsonConfigFileContent(this.config, ts.sys, inputPath, undefined, this.configFileName);
+  this.compile(config);
+  return Promise.resolve();
```
Comment (stefanpenner) [inline index.js]: if the compile is sync, no need in returning a promise

File: index.js [hunk 2]
```diff
@@ -10,50 +12,96 @@ var getCallerFile = require('get-caller-file');
...
+
+  var compilerHost = ts.createCompilerHost(config.options);
+
+  var inputPath = this.inputPaths[0];
+  var outputPath = this.outputPath;
+
+  var hostWriteFile = compilerHost.writeFile;
+  compilerHost.writeFile = function (fileName, data, writeByteOrderMark, onError) {
+    var outFileName = outputPath + fileName.substr(inputPath.length);
+    hostWriteFile(outFileName, data, writeByteOrderMark, onError);
```
Comment (stefanpenner) [inline index.js]: if nothing changed, do we still write a given file?

Comment (krisselden) [inline index.js]: taking your fix from rollup, I will add the async disk cache later.

Comment (stefanpenner) [inline index.js]: nice, does TS have decent caching already?

Comment (stefanpenner) [inline index.js]: I wonder if the `host` object can provide some protocol for detecting if something is stale or not. (looking to figure out a similar solution for sass)

**Pull request merged**